### PR TITLE
[issue 8337][Worker] Move initialize dlog namespace metadata to bin/pulsar

### LINKED ISSUE #8337: Create a tool to initialize dlog namespace
**Is your feature request related to a problem? Please describe.**

https://github.com/apache/pulsar/blob/master/pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java#L155

Currently, the starting function worker service will initialize the distributed log namespace. when initializing the distributed log namespace, the zookeeper will be connected. A better way is to implement a tool to initialize the distributed log namespace, which is similar to `./bin/pulsar initialize-cluster-metadata`

**Describe the solution you'd like**

* Implement a tool named `initialize-dlog-namespace-metadata`  to initialize the distributed log namespace
* Remove the initialization logic from the above code

**Describe alternatives you've considered**

* Add unit test
* Add document

**Additional context**

### PR DESCRIPTION
**Pull request by nodece**
Fixes #8337 

### Motivation

Currently, the starting function worker service will initialize the distributed log namespace. when initializing the distributed log namespace, the zookeeper will be connected. A better way is to implement a tool to initialize the distributed log namespace, which is similar to `./bin/pulsar initialize-cluster-metadata`.

### Modifications

- Add init distributed log namespace metadata to `./bin/pulsar initialize-cluster-metadata`
- Add flag to control whether the distributed log namespace is initialized in Function worker
- Add flag to `conf/functions_worker.yml`

### Doc

If you want to initialize distributed log metadata by `bin/pulsar`, you need to perform the following steps:
1. Using `./bin/pulsar initialize-cluster-metadata` to initialize cluster
2. Set initializedDlogMetadata to `true` in `functions_worker.yml`

**Note**: All the changes are compatible.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

### CODE REVIEW

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java [hunk 1]
```diff
@@ -150,15 +150,8 @@ private static URI initialize(WorkerConfig workerConfig)
...
-        // TODO: move this as part of pulsar cluster initialization later
-        try {
-            return WorkerUtils.initializeDlogNamespace(internalConf);
-        } catch (IOException ioe) {
-            log.error("Failed to initialize dlog namespace with zookeeper {} at metadata service uri {} for storing function packages",
-                internalConf.getZookeeperServers(), internalConf.getBookkeeperMetadataServiceUri(), ioe);
-            throw ioe;
-        }
+        // get the dlog namespace
+        return WorkerUtils.getDlogNamespaceURI(internalConf.getZookeeperServers());
```
Comment (sijie) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:154]: Don't remove the code immediately. Please add a flag in Function worker config to control whether do we need to initialize dlog namespace or not. By default, we can make it `true` (to initialize dlog namespace) for backward compatibility consideration and allow people to turn it off.

File: pulsar-broker/src/main/java/org/apache/pulsar/PulsarInitialDlogNamespaceMetadataSetup.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            System.exit(1);
+        }
+        if (arguments.bkMetadataServiceUri == null) {
+            System.err.println("Metadata service uri of BookKeeper argument is required (--bookkeeper-metadata-service-uri)");
+            jcommander.usage();
+            System.exit(1);
+        }
+
+        InternalConfigurationData internalConf = new InternalConfigurationData(arguments.configurationStore, arguments.configurationStore, null, arguments.bkMetadataServiceUri, null);
+        WorkerUtils.initializeDlogNamespace(internalConf);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/PulsarInitialDlogNamespaceMetadataSetup.java:66]: Also initialize the dlog namespace as part of `initialize-cluster-metadata` command.

**Review by sijie — CHANGES_REQUESTED** [review]
Overall the change looks good to me. I left two comments. PTAL.

**Review by zymap — APPROVED** [review]

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java (hunk 1 above)
Comment (nodece) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java]: I update my code, and I'm not sure if I'm doing the right thing, please let me know if have any problems.

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java [hunk 3]
```diff
@@ -59,6 +59,7 @@
     private GlobalZooKeeperCache globalZkCache;
     private ConfigurationCacheService configurationCacheService;
     private final ErrorNotifier errorNotifier;
+    private static boolean isInitialDlogMetadataByPulsarTool = false;
```
Comment (sijie) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:62]: How is this flag set?

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java [hunk 4]
```diff
@@ -82,7 +83,11 @@ protected void start() throws Exception {
         }
     }
 
-    private static URI initialize(WorkerConfig workerConfig)
+    public static void setEnableInitialDlogMetadataByPulsarTool(boolean enable) {
```
Comment (sijie) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:86]: How is this method used?

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java (hunk 3 above)
Comment (nodece) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:62]: @sijie Do you mean add a flag to config file?

Comment (sijie) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:62]: I mean how this flag is used. I don't see which method set the flag.

Comment (nodece) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:62]: https://github.com/apache/pulsar/pull/8781/files/4f5d8d5ce69a2b8f0efa2f4a72a39d5ca3cbc62a#diff-6ea96d1db19eb1c438ee988cc295057393319e0c0714771538e4ae9c7391cb86R87

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java (hunk 4 above)
Comment (nodece) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/Worker.java:86]: User can call `Woker. setEnableInitialDlogMetadataByPulsarTool(true)` or `Woker.setEnableInitialDlogMetadataByPulsarTool(false)`.

**Comment by tuteng** [comment]
ping @nodece

**Comment by nodece** [comment]
@tuteng @sijie could you review the PR?

**Comment by tuteng** [comment]
We may need to consider adding some documentation on how to use this command @Jennifer88huang

**Comment by Jennifer88huang** [comment]
@tuteng thank you for your reminder.
@nodece you can add related doc content in this PR or in a separate PR, or you can also ask a writer to add it. If you need any support, feel free to let us know. Thank you.

**Review by eolivelli — APPROVED** [review]
+1

**Comment by nodece** [comment]
@tuteng @Jennifer88huang thanks, I added a note, but I'm not sure when to add document to the official website.

**Comment by Jennifer88huang** [comment]
> @tuteng @Jennifer88huang once pr is merged, I will make a new PR for this doc.

That's great, thank you.

**Comment by nodece** [comment]
@tuteng @sijie tests has beed passed. could you merge this to master?

**Review by sijie — APPROVED** [review]

**Comment by nodece** [comment]
@Jennifer88huang I made a PR for document: https://github.com/apache/pulsar/pull/9185.

**Comment by codelipenghui** [comment]
Hi @tuteng I noticed this PR is based on #8560, #8560 has some big changes so it will not contains in the 2.7.x release. Do we need to cherry-pick this PR to branch-2.7?

**Comment by tuteng** [comment]
@codelipenghui Thanks, move this pr to 2.8.0, delete label v2.7.1

**Comment by sijia-w** [comment]
@nodece Just a soft reminder. We still need to add doc for the update in this pr.

> @tuteng @Jennifer88huang once pr is merged, I will make a new PR for this doc.

**Comment by Anonymitaet** [comment]
@sijia-w does the author already add docs in https://github.com/apache/pulsar/pull/9185?

**Comment by sijia-w** [comment]
@Anonymitaet Should I add the description in Functions Worker section? Thank you.

**Comment by sijia-w** [comment]
@Anonymitaet It seems so. Thanks for reminding me.

**Pull request merged**